### ISSUE TITLE
Rails 6 RC2 regression with multidb support

### ISSUE THREAD

**Issue by navied** [rails/rails#36757]
### Steps to reproduce

Git clone test repo: https://github.com/navied/autoloaderbug

```
rails db:create
rails db:migrate
rails s
```

### Expected behavior

Code should reload on code change in development without causing an error.

### Actual behavior

When a code change happens with autoloader set to classic the first hit to the application causes a NameError:

![image](https://user-images.githubusercontent.com/1896993/61837702-f2c49980-ae53-11e9-92f8-e71f7b2c09a3.png)

### System configuration
**Rails version**: 6.0.0 RC2

**Ruby version**: 2.5.5

**Comment by fxn** [comment]
I have been able to reproduce.

Problem is, there is a class object which is not assigned to any constant (`#<Class:0x00007fc694a12770>`), and whose name is `primary::SchemaMigration`. That is not a valid constant name.

This does not happen in `zeitwerk` mode because Zeitwerk assumes less things about the stuff it does not manage.

On a first thought, we could perhaps fix `classic` to not assume in `autoloaded?` the name of an arbitrary class passed as argument has to yield a valid constant name (unless anonymous). I'll work on it.

/cc @eileencodes (just FYI)

**Comment by shime** [comment]
I'm experiencing the same issue on `6.0.0` after updating from `5.2.3` when running `rake db:migrate`. 

This is happening when using either `:zeitwerk` or `:classic` as `config.autoloader`.

**Comment by OmriSama** [comment]
This seems to be supposedly fixed, but I'm trying to upgrade my Rails app to 6.0 right now and am experiencing a similar issue.

I have a method in a module that gets the DB Schema and returns it as a string. Calling it raises the `NameError: wrong constant name primary` exception:

![image](https://user-images.githubusercontent.com/10534779/66788566-e2451d80-ee9c-11e9-91a1-d0623f752650.png)

Interestingly, this is happening in one of my Specs, but not in `rails console`. Here's the apparent traceback:

![image](https://user-images.githubusercontent.com/10534779/66788700-6ac3be00-ee9d-11e9-9151-e5896b07fb78.png)

Marty is the name of the application.

**Comment by tconst** [comment]
@shime I was also having trouble with the `db:migrate` task.  In this case I tracked it down to a gem that was calling `self.class.model_name` during an `after_commit` callback on that `primary` named class. I would check to see if you have anything being included into ActiveRecord that does something similar.

https://github.com/betterup/wisper-activerecord-publisher/blob/master/lib/wisper/activerecord/publisher.rb#L46

**Comment by fxn** [comment]
I have tried the original application with the current stable version and cannot reproduce anymore.

I have also tried

```
 % bin/rails c
Loading development environment (Rails 6.0.2.2)
irb(main):001:0> ActiveRecord::Base.connection.migration_context.needs_migration?
   (0.7ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
=> false
```

and does not raise anything in both `classic` and `zeitwerk` modes.

Let's close then.

**Comment by alecslupu** [comment]
I am still having issues with this bug running rails 6.0.3

@fxn can you help me reopen the issue ? 

```rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

wrong constant name primary
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/inflector/methods.rb:282:in `const_get'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/inflector/methods.rb:282:in `block in constantize'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/inflector/methods.rb:280:in `each'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/inflector/methods.rb:280:in `inject'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/inflector/methods.rb:280:in `constantize'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/core_ext/module/introspection.rb:72:in `module_parents'
/lib/ruby/gems/2.5.0/gems/activemodel-6.0.3.1/lib/active_model/naming.rb:254:in `model_name'
/lib/ruby/gems/2.5.0/gems/wisper-activerecord-1.0.0/lib/wisper/active_record/publisher.rb:51:in `broadcast_model_name_key'
/lib/ruby/gems/2.5.0/gems/wisper-activerecord-1.0.0/lib/wisper/active_record/publisher.rb:43:in `after_commit_broadcast'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:428:in `block in make_lambda'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:273:in `block in simple'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:517:in `block in invoke_after'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:517:in `each'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:517:in `invoke_after'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:136:in `run_callbacks'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:825:in `_run_commit_callbacks'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/transactions.rb:339:in `committed!'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/connection_adapters/abstract/transaction.rb:127:in `commit_records'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/connection_adapters/abstract/transaction.rb:265:in `block in commit_transaction'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/connection_adapters/abstract/transaction.rb:255:in `commit_transaction'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/connection_adapters/abstract/transaction.rb:293:in `block in within_new_transaction'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/connection_adapters/abstract/transaction.rb:278:in `within_new_transaction'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/connection_adapters/abstract/database_statements.rb:280:in `transaction'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/transactions.rb:212:in `transaction'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/migration.rb:1361:in `ddl_transaction'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/migration.rb:1309:in `execute_migration_in_transaction'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/migration.rb:1281:in `block in migrate_without_lock'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/migration.rb:1280:in `each'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/migration.rb:1280:in `migrate_without_lock'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/migration.rb:1229:in `block in migrate'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/migration.rb:1382:in `with_advisory_lock'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/migration.rb:1229:in `migrate'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/migration.rb:1061:in `up'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/migration.rb:1036:in `migrate'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/tasks/database_tasks.rb:238:in `migrate'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/railties/databases.rake:86:in `block (3 levels) in <top (required)>'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/railties/databases.rake:84:in `each'
/lib/ruby/gems/2.5.0/gems/activerecord-6.0.3.1/lib/active_record/railties/databases.rake:84:in `block (2 levels) in <top (required)>'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/task.rb:273:in `block in execute'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/task.rb:273:in `each'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/task.rb:273:in `execute'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/task.rb:214:in `block in invoke_with_call_chain'
/lib/ruby/2.5.0/monitor.rb:226:in `mon_synchronize'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/task.rb:194:in `invoke_with_call_chain'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/task.rb:183:in `invoke'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/application.rb:160:in `invoke_task'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/application.rb:116:in `each'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/application.rb:116:in `block in top_level'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/application.rb:125:in `run_with_threads'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/application.rb:110:in `top_level'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/application.rb:83:in `block in run'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/application.rb:186:in `standard_exception_handling'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/lib/rake/application.rb:80:in `run'
/lib/ruby/gems/2.5.0/gems/rake-12.3.3/exe/rake:27:in `<top (required)>'
/bin/rake:23:in `load'
/bin/rake:23:in `<main>'
```

**Comment by shime** [comment]
Thanks for pointing me in the right direction, @tconst. For posterity, here's what happened in my project.

Our Gemfile contained [wisper-activerecord](https://github.com/krisleech/wisper-activerecord) which caused problems. Not sure how, but that gem clashes with Zeitwerk. 

The solution I used was to patch Active Support's Inflector to skip trying to constantize `"primary"`. 

In the meantime we have removed Wisper from the Gemfile and we no longer experience issues. This is indeed not an issue with Zeitwerk, but with Wisper.

Good call on closing this and thank you for your hard work on the new loader, @fxn.

**Comment by alecslupu** [comment]
I can confirm @shime statement. The wisper active record causes the issue. 

A fix has been proposed in https://github.com/krisleech/wisper-activerecord/pull/30

**Comment by fxn** [comment]
Can you help me understand the root cause of this?

Given a vanilla `rails new` application that has `wisper-activerecord` in the Gemfile, which would be a minimal change I can do to trigger the error?

**Comment by alecslupu** [comment]
@fxn , I have managed to replicate several times the issue having a migration like : 
```
class ConvertPersonToActiveStorage < ActiveRecord::Migration[6.0]
  def up
    Person.where.not(image_file_name: nil).find_each do |entity|
      Some::BackGroundJob.perform_later(entity.id)
    end
  end

  def down
    raise ActiveRecord::IrreversibleMigration
  end
end
```

The issue resides in the fact that `wisper-activerecord` will trigger the `after_commit` hook at the end of the migration (as described in https://github.com/krisleech/wisper-activerecord/blob/master/lib/wisper/active_record/publisher.rb#L41). 

The culprit of the error is the below function:  
```
      def broadcast_model_name_key
        self.class.model_name.param_key rescue false
      end
```
And the values, while describing the issue: 

```
[3] pry(#<primary::SchemaMigration>)> self.class
=> #<Class:0x0000557c6257c360>(version: string)
[4] pry(#<primary::SchemaMigration>)> self.class.model_name
NameError: wrong constant name primary
from /home/alecslupu/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-6.0.3.1/lib/active_support/inflector/methods.rb:282:in `const_get'
[5] pry(#<primary::SchemaMigration>)> 
```

If additional details are required, let us know

**Comment by fxn** [comment]
Aaahhh, I have reproduced. It is enough to throw

```ruby
Wisper::ActiveRecord.extend_all
```

in an initializer.

With that setting, Wisper is involved in migrations because `ActiveRecord::SchemaMigration` is an internal Rails model used to manage the table `schema_migrations`, and therefore when AR tries to write to that table. Is in those models that Wisper errs, but I believe that is not Wisper's fault.

The name of the schema migration class [is decorated in an unusual way](https://github.com/rails/rails/blob/dc715dd803a9d9152391644709d0ca1c420164e4/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L130-L146).

So the root problem here is:

* Multi database support creates a class object that is assigned to no constant (via `Class.new`), but overrides `name` and `to_s` that way so that they look like `primary::SchemaMigration`.
*  Wisper [calls `self.class.model_name`](https://github.com/krisleech/wisper-activerecord/blob/8610a4502abd037f59eda738d986c1dd444acbb6/lib/wisper/active_record/publisher.rb#L43) in a callback. That is a legit call.
* `AMo::Naming#model_name` in turn calls [`Module#module_parents`](https://github.com/rails/rails/blob/358ac36edf1695fcbec0aa21f126a3d8b83d4b5a/activesupport/lib/active_support/core_ext/module/introspection.rb#L67), which performs a `constantize` call.

Bottom line, this does not seem related to autoloading. The class objects created to manage to `schema_migrations` do not play well with the Active Module naming interface.

I am not sure right now about which is the best way to address this. @eileencodes do you have any thoughts about this? Why are those classes defined that way? Could we make them conform to the AMo interface?

**Comment by fxn** [comment]
Minimal way to reproduce in a vanilla Rails application:

1. `rails new foo`.
2. `rails g migration bar`.
3. Write `schema_migration.model_name` in the `change` method.
4. `rails db:migrate`.

You'll get

```
NameError: wrong constant name primary
```

Reason is the name of the class is "primary::SchemaMigration" and therefore `model_name` cannot compute `schema_migration.module_parents`.

**Comment by eileencodes** [comment]
I opened https://github.com/rails/rails/pull/39500, I think that will fix 6.0.

**Comment by byroot** [comment]
I believe this is fixed.
